File: trunk/esper/src/test/java/net/esper/regression/eql/TestPerf3StreamOuterJoinCoercion.java
package net.esper.regression.eql;

import junit.framework.TestCase;
import net.esper.client.EPServiceProvider;
import net.esper.client.EPServiceProviderManager;
import net.esper.client.EPStatement;
import net.esper.support.util.SupportUpdateListener;
import net.esper.support.bean.SupportBean;
import net.esper.event.EventBean;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

public class TestPerf3StreamOuterJoinCoercion extends TestCase
{
    private EPServiceProvider epService;
    private SupportUpdateListener listener;

    public void setUp()
    {
        epService = EPServiceProviderManager.getDefaultProvider();
        epService.initialize();
        listener = new SupportUpdateListener();
    }

    public void testPerfCoercion3waySceneOne()
    {
        String stmtText = "select s1.intBoxed as v1, s2.longBoxed as v2, s3.doubleBoxed as v3 from " +
                SupportBean.class.getName() + "(string='A').win:length(1000000) s1 " +
                " left outer join " +
                SupportBean.class.getName() + "(string='B').win:length(1000000) s2 on s1.intBoxed=s2.longBoxed " +
                " left outer join " +
                SupportBean.class.getName() + "(string='C').win:length(1000000) s3 on s1.intBoxed=s3.doubleBoxed";

        EPStatement stmt = epService.getEPAdministrator().createEQL(stmtText);
        stmt.addListener(listener);

        // preload
        for (int i = 0; i < 10000; i++)
        {
            sendEvent("B", 0, i, 0);
            sendEvent("C", 0, 0, i);
        }

        long startTime = System.currentTimeMillis();
        for (int i = 0; i < 5000; i++)
        {
            int index = 5000 + i % 1000;
            sendEvent("A", index, 0, 0);
            EventBean event = listener.assertOneGetNewAndReset();
            assertEquals(index, event.get("v1"));
            assertEquals((long)index, event.get("v2"));
            assertEquals((double)index, event.get("v3"));
        }
        long endTime = System.currentTimeMillis();
        long delta = endTime - startTime;

        assertTrue("Failed perf test, delta=" + delta, delta < 1000);
    }

    public void testPerfCoercion3waySceneTwo()
    {
        String stmtText = "select s1.intBoxed as v1, s2.longBoxed as v2, s3.doubleBoxed as v3 from " +
                SupportBean.class.getName() + "(string='A').win:length(1000000) s1 " +
                " left outer join " +
                SupportBean.class.getName() + "(string='B').win:length(1000000) s2 on s1.intBoxed=s2.longBoxed " +
                " left outer join " +
                SupportBean.class.getName() + "(string='C').win:length(1000000) s3 on s1.intBoxed=s3.doubleBoxed";

        EPStatement stmt = epService.getEPAdministrator().createEQL(stmtText);
        stmt.addListener(listener);

        // preload
        for (int i = 0; i < 10000; i++)
        {
            sendEvent("B", 0, i, 0);
            sendEvent("A", i, 0, 0);
        }

        listener.reset();
        long startTime = System.currentTimeMillis();
        for (int i = 0; i < 5000; i++)
        {
            int index = 5000 + i % 1000;
            sendEvent("C", 0, 0, index);
            EventBean event = listener.assertOneGetNewAndReset();
            assertEquals(index, event.get("v1"));
            assertEquals((long)index, event.get("v2"));
            assertEquals((double)index, event.get("v3"));
        }
        long endTime = System.currentTimeMillis();
        long delta = endTime - startTime;

        assertTrue("Failed perf test, delta=" + delta, delta < 1000);
    }

    public void testPerfCoercion3waySceneThree()
    {
        String stmtText = "select s1.intBoxed as v1, s2.longBoxed as v2, s3.doubleBoxed as v3 from " +
                SupportBean.class.getName() + "(string='A').win:length(1000000) s1 " +
                " left outer join " +
                SupportBean.class.getName() + "(string='B').win:length(1000000) s2 on s1.intBoxed=s2.longBoxed " +
                " left outer join " +
                SupportBean.class.getName() + "(string='C').win:length(1000000) s3 on s1.intBoxed=s3.doubleBoxed";

        EPStatement stmt = epService.getEPAdministrator().createEQL(stmtText);
        stmt.addListener(listener);

        // preload
        for (int i = 0; i < 10000; i++)
        {
            sendEvent("A", i, 0, 0);
            sendEvent("C", 0, 0, i);
        }

        listener.reset();
        long startTime = System.currentTimeMillis();
        for (int i = 0; i < 5000; i++)
        {
            int index = 5000 + i % 1000;
            sendEvent("B", 0, index, 0);
            EventBean event = listener.assertOneGetNewAndReset();
            assertEquals(index, event.get("v1"));
            assertEquals((long)index, event.get("v2"));
            assertEquals((double)index, event.get("v3"));
        }
        long endTime = System.currentTimeMillis();
        long delta = endTime - startTime;

        assertTrue("Failed perf test, delta=" + delta, delta < 1000);
    }

    private void sendEvent(String string, int intBoxed, long longBoxed, double doubleBoxed)
    {
        SupportBean bean = new SupportBean();
        bean.setString(string);
        bean.setIntBoxed(intBoxed);
        bean.setLongBoxed(longBoxed);
        bean.setDoubleBoxed(doubleBoxed);
        epService.getEPRuntime().sendEvent(bean);
    }

    private static final Log log = LogFactory.getLog(TestPerf3StreamOuterJoinCoercion.class);
}

File: NEsper.Tests/regression/eql/TestPerf3StreamOuterJoinCoercion.cs
// ---------------------------------------------------------------------------------- /
// Copyright (C) 2007 Esper Team. All rights reserved.                                /
// http://esper.codehaus.org                                                          /
// ---------------------------------------------------------------------------------- /
// The software in this package is published under the terms of the GPL license       /
// a copy of which has been included with this distribution in the license.txt file.  /
// ---------------------------------------------------------------------------------- /

using System;

using NUnit.Framework;

using net.esper.client;
using net.esper.compat;
using net.esper.events;
using net.esper.support.bean;
using net.esper.support.util;

using org.apache.commons.logging;

namespace net.esper.regression.eql
{
	[TestFixture]
	public class TestPerf3StreamOuterJoinCoercion
	{
	    private EPServiceProvider epService;
	    private SupportUpdateListener listener;

	    [SetUp]
	    public void SetUp()
	    {
            PropertyResolutionStyleHelper.DefaultPropertyResolutionStyle = PropertyResolutionStyle.CASE_INSENSITIVE;
            
            epService = EPServiceProviderManager.GetDefaultProvider();
	        epService.Initialize();
	        listener = new SupportUpdateListener();
	    }

	    [Test]
	    public void testPerfCoercion3waySceneOne()
	    {
	        String stmtText = "select s1.intBoxed as v1, s2.longBoxed as v2, s3.doubleBoxed as v3 from " +
	                typeof(SupportBean).FullName + "(string='A').win:length(1000000) s1 " +
	                " left outer join " +
	                typeof(SupportBean).FullName + "(string='B').win:length(1000000) s2 on s1.intBoxed=s2.longBoxed " +
	                " left outer join " +
	                typeof(SupportBean).FullName + "(string='C').win:length(1000000) s3 on s1.intBoxed=s3.doubleBoxed";

	        EPStatement stmt = epService.EPAdministrator.CreateEQL(stmtText);
	        stmt.AddListener(listener);

	        // preload
	        for (int i = 0; i < 10000; i++)
	        {
	            SendEvent("B", 0, i, 0);
	            SendEvent("C", 0, 0, i);
	        }

	        long startTime = DateTimeHelper.CurrentTimeMillis;
	        for (int i = 0; i < 5000; i++)
	        {
	            int index = 5000 + i % 1000;
	            SendEvent("A", index, 0, 0);
	            EventBean _event = listener.AssertOneGetNewAndReset();
	            Assert.AreEqual(index, _event["v1"]);
	            Assert.AreEqual((long)index, _event["v2"]);
	            Assert.AreEqual((double)index, _event["v3"]);
	        }
	        long endTime = DateTimeHelper.CurrentTimeMillis;
	        long delta = endTime - startTime;

	        Assert.IsTrue(delta < 1000,"Failed perf test, delta=" + delta);
	    }

	    [Test]
	    public void testPerfCoercion3waySceneTwo()
	    {
	        String stmtText = "select s1.intBoxed as v1, s2.longBoxed as v2, s3.doubleBoxed as v3 from " +
	                typeof(SupportBean).FullName + "(string='A').win:length(1000000) s1 " +
	                " left outer join " +
	                typeof(SupportBean).FullName + "(string='B').win:length(1000000) s2 on s1.intBoxed=s2.longBoxed " +
	                " left outer join " +
	                typeof(SupportBean).FullName + "(string='C').win:length(1000000) s3 on s1.intBoxed=s3.doubleBoxed";

	        EPStatement stmt = epService.EPAdministrator.CreateEQL(stmtText);
	        stmt.AddListener(listener);

	        // preload
	        for (int i = 0; i < 10000; i++)
	        {
	            SendEvent("B", 0, i, 0);
	            SendEvent("A", i, 0, 0);
	        }

	        listener.Reset();
	        long startTime = DateTimeHelper.CurrentTimeMillis;
	        for (int i = 0; i < 5000; i++)
	        {
	            int index = 5000 + i % 1000;
	            SendEvent("C", 0, 0, index);
	            EventBean _event = listener.AssertOneGetNewAndReset();
	            Assert.AreEqual(index, _event["v1"]);
	            Assert.AreEqual((long)index, _event["v2"]);
	            Assert.AreEqual((double)index, _event["v3"]);
	        }
	        long endTime = DateTimeHelper.CurrentTimeMillis;
	        long delta = endTime - startTime;

	        Assert.IsTrue(delta < 1000,"Failed perf test, delta=" + delta);
	    }

	    [Test]
	    public void testPerfCoercion3waySceneThree()
	    {
	        String stmtText = "select s1.intBoxed as v1, s2.longBoxed as v2, s3.doubleBoxed as v3 from " +
	                typeof(SupportBean).FullName + "(string='A').win:length(1000000) s1 " +
	                " left outer join " +
	                typeof(SupportBean).FullName + "(string='B').win:length(1000000) s2 on s1.intBoxed=s2.longBoxed " +
	                " left outer join " +
	                typeof(SupportBean).FullName + "(string='C').win:length(1000000) s3 on s1.intBoxed=s3.doubleBoxed";

	        EPStatement stmt = epService.EPAdministrator.CreateEQL(stmtText);
	        stmt.AddListener(listener);

	        // preload
	        for (int i = 0; i < 10000; i++)
	        {
	            SendEvent("A", i, 0, 0);
	            SendEvent("C", 0, 0, i);
	        }

	        listener.Reset();
	        long startTime = DateTimeHelper.CurrentTimeMillis;
	        for (int i = 0; i < 5000; i++)
	        {
	            int index = 5000 + i % 1000;
	            SendEvent("B", 0, index, 0);
	            EventBean _event = listener.AssertOneGetNewAndReset();
	            Assert.AreEqual(index, _event["v1"]);
	            Assert.AreEqual((long)index, _event["v2"]);
	            Assert.AreEqual((double)index, _event["v3"]);
	        }
	        long endTime = DateTimeHelper.CurrentTimeMillis;
	        long delta = endTime - startTime;

	        Assert.IsTrue(delta < 1000,"Failed perf test, delta=" + delta);
	    }

	    private void SendEvent(String _string, int intBoxed, long longBoxed, double doubleBoxed)
	    {
	        SupportBean bean = new SupportBean();
	        bean.SetString(_string);
	        bean.SetIntBoxed(intBoxed);
	        bean.SetLongBoxed(longBoxed);
	        bean.SetDoubleBoxed(doubleBoxed);
	        epService.EPRuntime.SendEvent(bean);
	    }

        private static Log log = LogFactory.GetLog(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);
	}
} // End of namespace
